### PR TITLE
Created participant event registration and created Address class

### PR DESCRIPTION
**Pull request by ghost**
(dtos,entities,services,controller for Address, repositories and exceptions).

### CODE REVIEW

File: src/main/java/ro/itschool/ema/controllers/ParticipantController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    @PostMapping("/events/{eventId}/participants")
+    public ResponseEntity<ParticipantDTO> createParticipant(@RequestBody @Valid ParticipantDTO participantDTO) {
+        return ResponseEntity.ok(participantService.createParticipant(participantDTO));
+    }
+
+
+
+
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/controllers/ParticipantController.java]: abide by the coding guidelines and remove unnecessary empty lines

File: src/main/java/ro/itschool/ema/exceptions/ExceptionHandlerAdvice.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+import java.util.Map;
+
+@Slf4j
+@ControllerAdvice
+public class ExceptionHandlerAdvice {
+
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/exceptions/ExceptionHandlerAdvice.java]: the exception handlers from this class need to be moved to the GlobalExceptionHandler and modified accordingly. If it is not clear how, merge to main as it is and I will do it afterewards.

File: src/main/java/ro/itschool/ema/models/entities/Address.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
+package ro.itschool.ema.models.entities;
+
+import lombok.Data;
+import jakarta.persistence.*;
+@Data
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Address.java]: This is not recommended on entities. remove it and use other annotations

File: src/main/java/ro/itschool/ema/models/entities/Address.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+@Entity
+@Table(name = "addresses")
+public class Address {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
+
+    @Column(name = "street")
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Address.java:21]: naming the entity column is not required if it is the same as the variable name. the only place you need it is on postal_code

File: src/main/java/ro/itschool/ema/services/AddressService.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package ro.itschool.ema.services;
+
+import ro.itschool.ema.models.dtos.AddressDTO;
+
+public interface AddressService {
+
+    AddressDTO createAddress(AddressDTO addressDTO);
+
+
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/services/AddressService.java]: delete empty lines

File: src/main/java/ro/itschool/ema/models/entities/Staff.java [hunk 6]
```diff
@@ -27,6 +27,9 @@ public Staff(String firstName, String lastName, LocalDate dateOfBirth, String em
     @JoinColumn(name = "organizer")
     private Organizer organizer;
 
+    @OneToOne(mappedBy = "staff")
+    private Address address;
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Staff.java]: ....fetch = FetchType.LAZY
@JoinColumn(name = "address_id")

File: src/main/java/ro/itschool/ema/services/AddressServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import org.springframework.stereotype.Service;
+import ro.itschool.ema.models.dtos.AddressDTO;
+import ro.itschool.ema.models.entities.Address;
+import ro.itschool.ema.repositories.AddressRepository;
+
+@Slf4j
+@Service
+public class AddressServiceImpl implements AddressService{
+
+
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/services/AddressServiceImpl.java]: delete empty lines

File: src/main/java/ro/itschool/ema/models/entities/Organizer.java [hunk 8]
```diff
@@ -24,6 +24,9 @@ public class Organizer {
     @OneToMany(mappedBy = "organizer")
     private List<Event> eventList = new ArrayList<>();
 
+    @OneToOne(mappedBy = "organizer")
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Organizer.java]: ....fetch = FetchType.LAZY

@JoinColumn(name = "address_id")

File: src/main/java/ro/itschool/ema/models/entities/Participant.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    @Column(name = "participant_name")
+    private String name;
+
+    @Column(name = "email")
+    private String email;
+
+    @Column(name = "phone_number")
+    private int phoneNumber;
+
+    @OneToOne(mappedBy = "participant")
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Participant.java]: ....fetch = FetchType.LAZY
@JoinColumn(name = "address_id")

File: src/main/java/ro/itschool/ema/models/entities/Address.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    private String city;
+
+    @Column(name = "postal_code")
+    private int postalCode;
+
+    @Column(name = "country")
+    private String country;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "organizer")
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Address.java]: I would delete this and make the owner the Organizer.

File: src/main/java/ro/itschool/ema/models/entities/Address.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    private Organizer organizer;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "participant")
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Address.java]: I would delete this and make the owner the Organizer.

File: src/main/java/ro/itschool/ema/models/entities/Address.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    private Participant participant;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "staff")
```
Comment (Dragos-U) [inline src/main/java/ro/itschool/ema/models/entities/Address.java]: I would delete this and make the owner the Organizer.

**Review by Dragos-U — CHANGES_REQUESTED** [review]
good job.
after changing it, you can merge to main

File: src/main/java/ro/itschool/ema/controllers/ParticipantController.java (hunk 1 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/controllers/ParticipantController.java]: Solved.

File: src/main/java/ro/itschool/ema/exceptions/ExceptionHandlerAdvice.java (hunk 2 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/exceptions/ExceptionHandlerAdvice.java]: Solved.

File: src/main/java/ro/itschool/ema/models/entities/Address.java (hunk 3 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Address.java]: Solved for Address and Participant entity.

File: src/main/java/ro/itschool/ema/services/AddressService.java (hunk 5 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/services/AddressService.java]: Solved.

File: src/main/java/ro/itschool/ema/services/AddressServiceImpl.java (hunk 7 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/services/AddressServiceImpl.java]: Solved.

File: src/main/java/ro/itschool/ema/models/entities/Address.java (hunk 4 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Address.java:21]: I am unsure how to handle this.
Wouldn't the organizer and staff need to see the participant's street, city and country?
To know where they are coming from and for security reasons?

File: src/main/java/ro/itschool/ema/models/entities/Staff.java (hunk 6 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Staff.java]: Solved.

File: src/main/java/ro/itschool/ema/models/entities/Organizer.java (hunk 8 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Organizer.java]: Solved.

File: src/main/java/ro/itschool/ema/models/entities/Participant.java (hunk 9 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Participant.java]: Solved.

File: src/main/java/ro/itschool/ema/models/entities/Address.java (hunk 10 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Address.java]: Solved.

File: src/main/java/ro/itschool/ema/models/entities/Address.java (hunk 11 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Address.java]: Solved.

File: src/main/java/ro/itschool/ema/models/entities/Address.java (hunk 12 above)
Comment (ghost) [reply src/main/java/ro/itschool/ema/models/entities/Address.java]: Solved.

**Pull request merged**